### ISSUE TITLE
Possible keymap improvements for Pro¹

### ISSUE THREAD

**Issue by Kabouik** [droidian/meta#7]
**tl;dr**
Make both ellow arrow keys (hereafter referred to as <kbd>Yellow</kbd>) a true modifier because everyone needs more modifiers in Linux, because thumb keyboards need modifier keys like that one that are replicated on both sides. Move or duplicate <kbd>/</kbd> (and hence <kbd>?</kbd>) to the first layer of another key because <kbd>/</kbd> belongs to the first layer on any keyboard and is probably one of the most used non-alpha keys.

---

With the kernel Droidian uses, the <kbd>Yellow</kbd> keys on the Pro¹ do not really work as modifiers because they have no keycode assigned. Somehow, they can still be chorded with all keys with a yellow print to produce that yellow character but I am not sure know how this works without a modifier keycode.

This is what we get with `libinput-tools` (`libinput debug-events):

```
Ctrl: *** (-1) (same as many other keys, including Space, Ctrl, Shift, </>, and alpha keys)
**Yellow arrow: nothing**
Fx: KEY_LEFTMETA (125)
Alt: *** (-1) 
Sym: KEY_RIGHTALT (100)
```

If we start an X window with Xephyr and check keycodes therein with `xev`, er get this:

```
Ctrl: Control_L (keycode 37)
**Yellow arrow: nothing**
Fx: Super_L (keycode 133)
Alt: Alt_L (keycode 64)
Sym: ISO_Level3_Shift (keycode 108)
```

I see two valid reasons for having those <kbd>Yellow</kbd> keys produce the secondary layer yellow characters:
- this is how the keycaps are printed, and hence is easily discoverable and consistent
- this is what stock Android does (probably for the same reason: following the key cap printings)

However, there are also significant downsides:
- this is fully redundant with <kbd>Shift</kbd>, since all characters printed in yellow on the keyboard are secondary layer characters (except <kbd>/</kbd> and <kbd>?</kbd>, but see below) and the standard modifier for the secondary layer is <kbd>Shift</kbd>, which Droidian does also respect
- we lose something really precious for Mobile Linux and a thumb keyboard: a modifier key that is replicated on both sides, which can therefore be combined with both thumbs and therefore with literally any key on the keyboard without weird moves
- without that, we effectively have just <kbd>Shift</kbd> and <kbd>Ctrl</kbd> available for both thumbs at the moment, and those cannot be combined with alphanumeric keys to produce other characters; also <kbd>Shift</kbd> produces uppercase, so that leaves us with just <kbd>Ctrl</kbd>, which is rarely available for custom keybinds because it's already used with most keys for either system-wide or application-wide bindings; meaning very little space for keybinding customization
- the keyboard is hardly usable for international languages and keymaps (such as `us-intl`) because `ISO_Level3_Shift`, the modifier that triggers third-layer characters, is only on the left (<kbd>SYM</kbd>)
- since the choice is made to follow the key prints, <kbd>/</kbd> and <kbd>?</kbd> end up on a very unusual placement, on separate keys, and require using a key combination despite being very commonly used; of course <kbd>?</kbd> requires a key combination on standard keyboards too, but with the standard modifier (<kbd>Shift</kbd>), not with some weird key that is actually required only for those two characters across the whole key map (all other characters can be equally produced with <kbd>Shift</kbd>), which really breaks muscle memory

The folks who ported Sailfish to the Pro1 went with a bold move and decided not to follow the key prints:
- <kbd>SYM</kbd> becomes a dedicated key for <kbd>/</kbd>/<kbd>?</kbd> and put thems back together
- even though the placement of the key is surprising, it at least mobilises the same muscle memory as a standard keyboard because the characters are readily accessible without a unique modifier reserved to those only two characters
- being freed from having <kbd>Yellow</kbd> used to produce two regular characters that normally sit on their own key, they could make it `ISO_Level3_Shift` (= `Alt_R`), which makes it ideal for international keymaps and third-layer characters across the whole keyboard without painful thumb stunts
- for those who don't need `Alt_R` for international characters or who are ready to sacrifice it (by using dead keys or a compose key for instance), this gives a fully-customizable modifier that can be used with both thumbs for dozens of new keybindings (very useful in tiling WMs for instance)

In Sailfish, this is what we get with `evdev_trace -t`:

```
Ctrl: LEFTCTRL
**Yellow arrow: RIGHTALT**
Fx: LEFTMETA
Alt: LEFTALT
Sym: Slash
```

The patches they used are [here](https://github.com/sailfish-on-fxtecpro1/kernel-fxtec-pro1/commit/ca864049bb50eb2c736aa318b60b487e437b25ca) and [here](https://github.com/sailfish-on-fxtecpro1/kernel-fxtec-pro1/commit/c1635969787510687253c4fd860ac420e99e181a). I am not sure that they can be applied directly on what Droidian uses, I think @r3vn tried and got issues. Also, in Sailfish, I fail to distinguish `LEFTALT` and `RIGHTALT`, they all just behave as `ALT` it seems, but this may very likely be a Sailfish limitation that would not exist in Debian.

I realise these changes may not suit everyone. Not everyone wants an extra keybinding, or an international language, or to learn a keymap that does not 100% resembles the keyprints, but in my opinion the advantages outweigh the downsides so this could be worth an opt-in branch.

Alternatively, <kbd>YELLOW</kbd> could be made `RIGHTALT`/`ISO_Level3_Shift` **and** yellow printed character could be replicated on both second and third layers of the keymap. This should result in the same default behavior as currently (<kbd>Shift</kbd> would still trigger secondary layer, and <kbd>Yellow</kbd> would trigger third layer, and those would be the same in the keymap), except users could then customize the keymap to their liking if needed, or use the new modifier for other uses. Same with <kbd>/</kbd>/<kbd>?</kbd>: this could be made a single key in place of <kbd>SYM</kbd> which has no real meaning outside Android, and is currently mapped to `ISO_Level3_Shift` (but again, this modifier is most useful if replicated on both sides, which would be the case with the above changes by making. On a full size keyboard, this modifier is located on the right (<kbd>AltGr</kbd> on ISO keyboards), but on a thumb keyboard it is hardly usable with all alpha keys if located on the left only.

Sorry for the lengthy post, the **tl:dr** above should be enough, but since this is a fairly subjective suggestion, I felt it was best to elaborate and give real-world examples.

**Comment by HookPro1** [comment]
I agree that getting the question mark and the slash on the same key without the use of the arrow key (Fn) would be very important.  Both are very common and the placement of the Fn key makes it much more awkward than the shift key,  I don't care how it's accomplished and prints be damned, this would  be a great thing for anyone using to Pro1 to write.

**Comment by Kabouik** [comment]
Just for future reference or testing, I'm uploading here the `boot.img` built from https://github.com/droidian-devices/linux-android-fxtec-pro1/pull/2 because I'm having trouble building it now that the porting guide uses Trixie. No one has to trust that file of course and it will always be best to rebuild from sources, but it'll at least help me in the future if I need it and still can't rebuild.

<img width="979" height="217" alt="Image" src="https://github.com/user-attachments/assets/37d6fc1c-2880-4a3b-86c5-24c50f6ac3fc" />

[boot.img.tar.gz](https://github.com/user-attachments/files/21919045/boot.img.tar.gz), `sha256sum` of uncompressed `boot.img`: `85cea451eec057fa7e734548ca3ba6d779ed5836a3f9de14b8394575ef0d7d8e`

**Comment by g7** [comment]
Please also try

```
deb http://droidian-devices-linux-android-fxtec-pro1.repo.droidian.org/next-pr2/ trixie main
```

which is the same thing, but easier to install 😉
